PukingPredator/Assets/Scripts/Inventory.cs

```diff
@@ -45,45 +45,113 @@ public int itemCount
     /// Triggered whenever items enter/leave the inventory or decay. This will
     /// specifically trigger after the change has taken effect.
     /// </summary>
-    //TODO: make this trigger when the items decay
     public event Action onChange;
 
+    /// <summary>
+    /// The object that owns the inventory, ie the player.
+    /// </summary>
+    public GameObject owner;
+
     /// <summary>
     /// The mass of all items in the inventory.
     /// </summary>
     public float totalMass => items.Select(i => i.mass).Sum();
 
 
 
+    /// <summary>
+    /// Checks if the inventory contains the given item.
+    /// </summary>
+    /// <param name="item"></param>
+    /// <returns></returns>
+    public bool ContainsItem(Consumable item)
+    {
+        return items.Contains(item);
+    }
+
+    /// <summary>
+    /// Handles behaviour for whenever an item is added.
+    /// </summary>
+    /// <param name="item"></param>
+    private void ItemAdded(Consumable item)
+    {
+        item.inventory = this;
+        onChange?.Invoke();
+    }
+
+    /// <summary>
+    /// Handles behaviour for whenever an item is removed.
+    /// </summary>
+    /// <param name="item"></param>
+    private void ItemRemoved(Consumable item)
+    {
+        item.inventory = null;
+        onChange?.Invoke();
+    }
+
     /// <summary>
     /// Add an item to the inventory
     /// </summary>
     /// <param name="item">The item to be added.</param>
-    public void PushItem(Item item)
+    public void PushItem(Consumable item)
     {
         items.Insert(0, item);
```

Comment:
does inserting at 0 cause the list to copy itself? Would appending/removing to the end of the list be more efficient?

Reply:
i dont know the details of C#s implementation of lists enough to know how much it matters in general, but in our case we are doing this so few times that its not worth thinking about. optimize when necessary, doing so early on is a waste of resources unless its obvious that it will be an issue down the road